### ISSUE TITLE
Resend MFA Code API  

### ISSUE THREAD

**Issue by aoloo** [aws-amplify/amplify-js#6676]
**Is your feature request related to a problem? Please describe.**
Our workflow MFA is required in Cognito via (SMS). The user enters the login username and password and is redirected to verify the MFA code page. Now comes the edge case where a user does not receive the MFA code due to network issues or other interference. Therefore we need to provide users with the option to resend an MFA code.

**Describe the solution you'd like**
A method to resend MFA Code due to some edge case a user does not receive a code within the initial sign In flow. 
A similar hook to `Auth.resendSignUp()` but for MFA Code. 

**Describe alternatives you've considered**
Calling the `Auth.SignIn()` to resend MFA Code is not ideal because signIn requires a username and password.

**Comment by elorzafe** [comment]
@aoloo 

doing `Auth.signIn(..)` will resend the code, in that case you will need to store in memory the username and password for that purpose.

**Comment by aoloo** [comment]
@elorzafe @amhinson  we are currently implementing that, but the authentication code does not always get sent? There some instances where calling `Auth.signIn()` does not work.
```
try {
  const user = await Auth.signIn(username,password);
} catch (err) {
   `Code was not sent successfully!`
 }
```

This is my current implementation within a on click event when a user submits a form. It will be really nice to have a resendMFA api call. This can benefit the community implementing MFA.

**Comment by aoloo** [comment]
@elorzafe Just to note when calling `Auth.signIn` it returns `{"__type":"CodeMismatchException","message":"Invalid code or auth state for the user."}`. So it is crucial that we have an api for re sending mfa authentication code.

**Comment by mitchgillin** [comment]
Seconding! This is an issue for me as well. Would be great if I could resend confirmation codes, without having to make users re-sign in.

**Comment by pinpointpanda** [comment]
Yeah, also seeing this. We'll workaround with posting to `Auth.logIn` again, but it does feel like there ought to be an api method for this - it's the only part of the sign up/sign in/reset password etc... workflows that doesn't have a re-send.

**Comment by aoloo** [comment]
@harrysolovay @elorzafe any movement on this?

**Comment by MaxwellOldshein** [comment]
Has anyone heard any update on this issue?

**Comment by martinlanglois** [comment]
Just want to add my name to the list of people who would need this feature. 
I am not using the Amplify framework though. Only the php SDK. But it's visibly kind of all the same !

**Comment by chaawlaapooja** [comment]
Just want to add my name to the list of people who require this feature.
This is a blocker for my project.

**Comment by Thomsen-c** [comment]
This is a blocker on one of my projects as well.

**Comment by pedrohff360** [comment]
Same here... there's some update?

**Comment by justinslalom** [comment]
+1, this is a needed feature

**Comment by sfratini** [comment]
+1 This is very basic feature request at this point. Specially since we would like to send a code to actually validate the MFA before actually enabling it, to confirm the user has access to the phone/email. 
AWS enabled it right away, without validating anything. And it is funny, since AWS actually ask you for TWO codes before enabling.

**Comment by tigrenok00** [comment]
+1

**Comment by eugendorin** [comment]
+1

**Comment by jwelfare** [comment]
+1

**Comment by 44mkashif** [comment]
Any update on this issue? This is a blocker on one of my projects as well.

**Comment by AugustDev** [comment]
Need update on this

**Comment by Levisnkyyyy** [comment]
Seeking update as well

**Comment by madlerpar** [comment]
Is there any update on this? The given "workaround" of using `Auth.SignIn()` does not seem to actually work.

**Comment by Muzammil98** [comment]
+1, this is a needed feature

**Comment by JJ810** [comment]
If anyone is still looking for this solution here is my workaround/solution.

So to resend OTP code you should do `Auth.signIn(email, password)` again and update the user object with new response and do `Auth.confirmSignIn(user, otp, user.challengename)` again.

Here is my code:

```js
...
const [user, setUser] = useState()
...
 const handleSubmitMfa = async () => {
    if (user) {
      if (user.challengeName === 'SMS_MFA' || user.challengeName === 'SOFTWARE_TOKEN_MFA') {
        try {
          await Auth.confirmSignIn(user, otp, user.challengeName)
          router.push('/')
        } catch (err: any) {
          if (err.code === 'CodeMismatchException') {
            setError('Invalid OTP Code')
          }
        }
      } else if (user.challengeName === 'MFA_SETUP') {
        Auth.setupTOTP(user)
      }
    }
  }

  const handleResend = async () => {
    if (user && user.challengeName === 'SMS_MFA') {
      const userRes = await Auth.signIn(values.email, values.password)
      // Update user object to confirm sign in with newly received code.
      setUser(userRes)
    }
  }
```

Hope it helps!

**Comment by LazyAfternoons** [comment]
The solution provided by @JJ810 seems to work just fine but we really need this feature.

**Comment by abdallahshaban557** [comment]
Hello everyone! Our team is working on introducing this as a feature of our library in a future iteration. We will provide an update on this Github issue when a new version is out. We do not yet have exact timelines, but we will share them once they are known to us!

**Comment by kceb** [comment]
I too need a solution that does not require me to cache the password... seems pretty insecure. Thankfully I'm doing this server-side. Still would prefer an API that would just resend the code

**Comment by ovidiu-a** [comment]
+1, this is a needed feature

**Comment by erinleigh90** [comment]
Cognito does not currently provide the ability for us to resend the MFA code. Until they provide this functionality, we cannot implement in amplify. We will continue to stress the importance of this issue internally and provide an update as soon as we have one.

**Comment by aoloo** [comment]
@abdallahshaban557  I would prefer we close #2010 and keep this one open. It is more active and recent. Thanks!

**Comment by abdallahshaban557** [comment]
@aoloo - that is true - I see that as well. Done!

**Comment by abduncan** [comment]
+1

**Comment by luigi1610** [comment]
the workaround of @JJ810 is not praticable. We cannot stor user password. This problem is present even with aws-sdk module, it seems that cognito miss this essential feature, How is it possible? Today no solution yet.

**Comment by abdallahshaban557** [comment]
We are working with the Cognito team to enable this feature soon. Once that feature is available in Cognito, the Amplify team is intending on introducing an API to handle this ASAP. We will provide updates on this GH issue.

**Comment by jmnProcom** [comment]
Also looking for this feature. But for now, is there a known workaround to trigger an SMS code when SMS is enabled but NOT the preferred method? Signing in the user again would not work in this case.

For instance, a user logging in with "SOFTWARE_TOKEN_MFA" set as preferred method but for some reason that user don't have access to his app and want to use SMS instead.

To be more specific, talking about the following: 
![image](https://github.com/aws-amplify/amplify-js/assets/95230152/5088cd34-4a56-4f3b-9e04-96a818c1b906)

**Comment by deepaktammali** [comment]
The work-around being discussed works for username/password login, Is there a way we can resend confirmation code for other providers such as google?

**Comment by abdallahshaban557** [comment]
Hello everyone, the Cognito team is collecting feedback since they are planning on enabling this feature. If you are open to having us reach out and get your feedback on our proposed experience, can you please send us an email to aws-amplify-customer@amazon.com with the subject of `Resend MFA feedback`

**Comment by ashwinchandran13** [comment]
I'm receiving the MFA code automatically but on using the SMS MFA codes I'm getting a 'CodeMismatchException: Invalid code or auth state for the user' just as @aoloo described

**Comment by ashwinchandran13** [comment]
### CodeMismatchException: Invalid code or auth state for the user

A closed [issue](https://github.com/aws-amplify/amplify-js/issues/3066) had a discussion what @aoloo described. Interestingly, the issue was closed without having a solution which rendered the discussion to be locked which lead me to posting the solution here but since anyone who is searching this issue is lead to here, here is the solution:

After receiving the code as SMS, **do not trigger/call authenticate flow once again to send the SMS MFA code** which causes the code to be invalidated and after getting the error you might notice another SMS MFA code being send which is due to the last authenticate call done to send initial SMS MFA code. 

Instead create another path or method in which you can use either `adminRespondToAuthChallenge` or `sendMFACode` provided the `Session` parameter must have a value of session from the last authenticate call.

In my case, I had built REST API's which calls the authenticate function within a Lambda. 

`authenticate` function is called after user submits email & password in frontend.

```js

    import { CognitoUser } from 'amazon-cognito-identity-js';

    // inside authenticate lambda function
    const userData = {
       Username: user.email,
       Pool:     userPool
    };
    const cognitoUser = new CognitoUser(userData);
    cognitoUser.authenticateUser(authenticationDetails, {
      onSuccess: ...,
      onFailure: ...,
      mfaRequired: (challengeName, challengeParameters) => {
        const { Session } = cognitoUser;
        if (payload.mfa === undefined) {
         // returns mfaRequired response to frontend
          resolve({mfaRequired: true, session: Session}) 
        }
      }
   })
```

`sendSMSMFA` Lambda function which is called after user enters the code into the frontend

```js
// inside sendSMSMFA function
const cognitoAdminUser = new CognitoIdentityServiceProvider({apiVersion: '2016-04-18'});
cognitoAdminUser.adminRespondToAuthChallenge({
 UserPoolId: user_pool_id,
 ClientId: client_id,
 ChallengeName: 'SMS_MFA',
 Session: payload.session,
 ChallengeResponses: {
   SMS_MFA_CODE: payload.mfaCode,
   USERNAME: payload.email
  },
 }, (err, data) => {
   if (err) {
    console.log('error responding to auth challenge', err);
    reject(err);
   } else {
      console.log('Successfully responded to auth challenge', data);
      const idToken = data.AuthenticationResult.IdToken;
      const AccessToken = data.AuthenticationResult.AccessToken;
      const refreshToken = data.AuthenticationResult.RefreshToken;
      resolve({token: idToken, access: AccessToken, refresh: refreshToken})
    }
   })
```

For anyone who wonders about why this isn't happening when using TOTP software token mehod is because the MFA code clock is separate of that of the cognito clock and thus there will always be a sync there even if codes are send in different `authenticate` calls but SMS MFA is really dependent on **session id**.

**Comment by felipecrestani** [comment]
??????

**Comment by valentinbeggi** [comment]
+1

**Comment by sagenate24** [comment]
+1

**Comment by elton-bz** [comment]
+1

**Comment by fcoradini** [comment]
+1

**Comment by Sc0ra** [comment]
+1

**Comment by owhittlef** [comment]
+1... blocker for my project. 

I will use the workaround for now, but as others have said, I do not want to cache the user's password.

**Comment by evan-macgregor** [comment]
We were promised this was being worked on a almost two years ago.

Most recent action I see is a closure of the related issue in aws-sdk-js as "not planned" in September 2023.

This really is an essential (and easy to implement) feature. Too bad Amazon is a struggling 2-trillion-dollar company and doesn't have the resources to deal with the 500 open issues on this library that's arguably necessary to use their services.

**Comment by rohith-dev10** [comment]
this feature is a must have. Any updates??

**Comment by ashwinchandran13** [comment]
@evan-macgregor @rohith-dev10 I implemented this by doing a reauthentication. Is that not possible in your cases?

**Comment by kennydifiore** [comment]
Reauthentication is a disruptive and undesirable user experience.  You would force the user to re-enter their credentials again.  For security purposes, the user's credentials should never be stored by the web application.  Every other IdP seems to support the ability to resend the MFA code because they offer API calls separate from Sign In/ Log In that allows the application to trigger a resend of the SMS MFA code.  This seems like an essential feature that should be implemented.  

Is the thought here that since SMS is the least secure of the MFA options that AWS would prefer us to migrate away from SMS and use TOTP apps or other alternatives?

**Comment by evan-macgregor** [comment]
@ashwinchandran13 Yeah, there's a workaround. That's not a feature or a fix, and as stated by many others, it's undesirable at best and insecure at worst. Don't pretend it's ok.

I, like many others, choose to use AWS because the marketing claims it is feature-complete. This is not the case, and many services offered by AWS are in such a state. You've wasted my time by letting me get this far into implementation, and if you decided that it's not that important, I guess I'll just have to change my implementation to use one of the many other companies that offer very similar features. Luckily I've written my code to easily strip out AWS, since I've recognized that this would likely happen.

**Comment by royjin-xcend** [comment]
It's incredibly disappointing. We chose Cognito based on AWS's branding, reputation, and the feature set promoted on their website. Resending a code is a basic, standard feature, and we didn't expect to encounter issues with it during the Cognito + Amply integration. We're seriously considering whether Cognito is the right fit for us moving forward. I hope other developers see this thread and avoid making the same mistake.

**Comment by rongxcend** [comment]
As an ex-architect from AWS, it is really shame seeing the Customer Obsession principle is deteriorating, given such a basic and essential feature does not raise any attention to AWS.  (4 years, what are you doing AWS?) 

Of course there will be smarter person from AWS using 'has backbone' leadership principal to write a very long tech paper to illustrate/prove why such feature might not be a good idea to be implemented.  
Nevertheless, it is an essential feature for a product claim to be IdP.

**Comment by mmartinez** [comment]
Is it possible to use:
https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GetUserAttributeVerificationCode.html ?

**Comment by MrHOY** [comment]
October 2024 and anything new?
Using a strict cache username and password and then calling API login again is not so good

**Comment by lpv90** [comment]
still no solution from AWS?

**Comment by Yoac-eng** [comment]
How can this still be open 4 years later, seems like theres nothing yet

**Comment by 117** [comment]
someone tag bezos at this point lmao

**Comment by LiliumYang** [comment]
Any update? Can we get an ETA for this feature request?

**Comment by sators** [comment]
Stardate 2025.4.08 — No communication from Starfleet Command. We await further instructions... and maybe a sign they're still out there.

**Comment by varmancholan** [comment]
Any updates?
